// @microsoft/spfx-heft-plugins/lib/spfxHeftPlugin.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const DevDeployPlugin_1 = __importDefault(require("./plugins/devDeployPlugin/DevDeployPlugin"));
const WebpackConfigurationPlugin_1 = __importDefault(require("./plugins/webpackConfigurationPlugin/WebpackConfigurationPlugin"));
const CustomizeWebpackConfigurationPlugin_1 = __importDefault(require("./plugins/CustomizeWebpackConfigurationPlugin"));
const PluginNames = __importStar(require("./plugins/PluginNames"));
const WebpackServeConfigurationPlugin_1 = __importDefault(require("./plugins/webpackConfigurationPlugin/WebpackServeConfigurationPlugin"));
/**
 * @public
 */
class SpfxHeftPlugin {
    constructor() {
        this.pluginName = PluginNames.SpfxHeftPlugin;
    }
    get accessor() {
        if (!this._accessor) {
            this._accessor = {
                devDeploy: DevDeployPlugin_1.default.accessor,
                spfxWebpackConfiguration: WebpackConfigurationPlugin_1.default.accessor,
                spfxWebpackServeConfiguration: WebpackServeConfigurationPlugin_1.default.accessor
            };
        }
        return this._accessor;
    }
    apply(heftSession, heftConfiguration) {
        DevDeployPlugin_1.default.apply(heftSession, heftConfiguration);
        WebpackConfigurationPlugin_1.default.apply(heftSession, heftConfiguration);
        WebpackServeConfigurationPlugin_1.default.apply(heftSession, heftConfiguration);
        CustomizeWebpackConfigurationPlugin_1.default.apply(heftSession, heftConfiguration, {
            customizationFilename: 'spfx-customize-webpack.js'
        });
    }
}
exports.default = new SpfxHeftPlugin();
